sklearn/tests/test_multiclass.py

```diff
@@ -705,6 +709,23 @@ def test_ecoc_float_y():
                          " got -1", ovo.fit, X, y)
 
 
+def test_ecoc_delegate_sparse_base_estimator():
+    # Non-regression test for
+    # https://github.com/scikit-learn/scikit-learn/issues/17218
+    check_func = functools.partial(check_array, accept_sparse=False)
+    base_est = CheckingClassifier(check_y=check_func, check_X=check_func)
+    ecoc = OutputCodeClassifier(base_est, random_state=0)
+    X, y = iris.data, iris.target
+    X_sp = sp.csc_matrix(X)
+    err_msg = 'A sparse matrix was passed, but dense data is required.*'
+    with pytest.raises(TypeError, match=err_msg):
+        ecoc.fit(X_sp, y)
```

Comment:
also fit with `X` and test `predict`?

Reply:
Im getting a weird `ValueError` when trying to first fit the estimator on non-sparse data before calling `predict`.
```
>           assert self.check_X(X, **params)
E           ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
It seems to be coming from inside CheckingClassifier's `fit` method. Maybe @glemaitre can help?

Reply:
I can call `ecoc.estimator.check_X(X)` just fine outside of the `fit` method though

Reply:
change `check_func` to return True, should fix it

Reply:
Wouldn't that mean there wont be any exception raised after calling `predict` since it internally uses the fitted `estimator_` instances of CheckingClassifier to make a prediction? Those fitted instances would still contain a modified `check_func`. Unless maybe i'm not understanding your suggestion fully. Here is what I tried:
```
  715     check_X = functools.partial(check_array, accept_sparse=False)
  716     check_y = functools.partial(check_array, accept_sparse=False, ensure_2d=False)
...
  727     with pytest.raises(TypeError, match=err_msg):
  728         ecoc.estimator.check_X = lambda x: True
  729         ecoc.estimator.check_y = lambda x: True
  730         ecoc.fit(X, y)
  731         for estimator in ecoc.estimators_:
  732             estimator.check_X = check_X
  733             estimator.check_y = check_y
  734         ecoc.predict(X_sp)
```
lines 728 and 729 bypass the `ValueError` but calling predict on a sparse matrix afterwards doesnt raise the expected exception and I get: 
```
>           ecoc.predict(X_sp)
E           Failed: DID NOT RAISE <class 'TypeError'>
```

Reply:
I mean something like

````py

def check_func():
	check_array....
	return True

```

Reply:
Thanks, the next problem im facing is that the predict method throws no `AssertionError`. The reason is that `OutputCodeClassifier` uses `_predict_binary` inside its `predict` method, and in turn `_predict_binary` uses CheckingClassifier's `decision_function` method since its implemented. The problem is  CheckingClassifier's decion_function does not perform array checks for binary classes so `check_func` is never called.
```
  715     def check_func(X):
  716         ensure_2d = False if X.ndim == 1 else True
  717         out = False
  718         try:
  719             check_array(X, accept_sparse=False, ensure_2d=ensure_2d)
  720             out = True
  721         except ValueError:
  722             pass
  723         finally:
  724             return out
  725 
  729     base_est = CheckingClassifier(check_y=check_func, check_X=check_func)
  730     ecoc = OutputCodeClassifier(base_est, random_state=0)
  731     X, y = iris.data, iris.target
  732     X_sp = sp.csc_matrix(X)
  733     assert_raises(AssertionError, ecoc.fit, X_sp, y)
  734 
  735     # test if exception is raised when calling `predict`
  736     ecoc.fit(X, y)
  737     ecoc.predict(X_sp)
```
`AssertionError: AssertionError not raised by predict`

Reply:
Yes I saw this issue when rebasing https://github.com/scikit-learn/scikit-learn/pull/15230
If we merge this PR, you will be fine because it imposes the check on X in `predict_proba` and `decision_function` as well.

Reply:
Alright, thanks. I'll wait for that PR then.